api/userpassloginprovider.go

```diff
@@ -45,26 +49,100 @@ func NewUserpassLoginProvider(
 	}
 }
 
+// tagToString returns the value of a tag's String method, or "" if the tag is nil.
+func tagToString(tag names.Tag) string {
+	if tag == nil {
+		return ""
+	}
+	return tag.String()
+}
+
 // userpassLoginProvider provides the default juju login provider that
 // authenticates the entity with the given name and password or macaroons. The
 // nonce should be empty unless logging in as a machine agent.
 type userpassLoginProvider struct {
-	tag          names.Tag
+	tag          string
```

Comment:
At this point "tag" is no longer a Tag but a string. Maybe call it "tagStr" or even "userInfo" since really, that's what it is. I don't think "user" is suitable because the format would be "user-fred".

Reply:
I'll go with `userInfo`.